overlord/registrystate/registrymgr.go

```diff
@@ -37,3 +49,114 @@ func setupRegistryHook(st *state.State, snapName, hookName string, ignoreError b
 	task := hookstate.HookTask(st, summary, hookSup, nil)
 	return task
 }
+
+type changeViewHandler struct {
+	hookstate.SnapHookHandler
+	ctx *hookstate.Context
+}
+
+func (h *changeViewHandler) Done() error {
+	h.ctx.Lock()
+	defer h.ctx.Unlock()
+
+	t, _ := h.ctx.Task()
+	tx, _, err := GetStoredTransaction(t)
+	if err != nil {
+		return fmt.Errorf("cannot get transaction in change-registry handler: %v", err)
+	}
+
+	if tx.aborted() {
+		return fmt.Errorf("cannot change registry %s/%s: %s rejected changes: %s", tx.RegistryAccount, tx.RegistryName, tx.abortingSnap, tx.abortReason)
+	}
+
+	return nil
+}
+
+type saveViewHandler struct {
+	hookstate.SnapHookHandler
+	ctx *hookstate.Context
+}
+
+func (h *saveViewHandler) Error(origErr error) (ignoreErr bool, err error) {
+	h.ctx.Lock()
+	defer h.ctx.Unlock()
+
+	t, _ := h.ctx.Task()
+	st := h.ctx.State()
+
+	var commitTaskID string
+	if err := t.Get("commit-task", &commitTaskID); err != nil {
+		return false, err
+	}
+
+	// stop all tasks (we need to do this manually because we're not erroring yet)
+	for _, t := range t.Change().Tasks() {
```

Comment:
> Could this use `t.Change().Abort()`?
AFAICT, no because it wouldn't start undoing tasks immediately. That would unset the ongoing transaction state so that could happen while we're rolling back, leading to concurrency issues (that behaviour isn't merged in master yet, but it's in the spike). The intent here was to stop everything while we roll back the save-views and then surface the error so the normal error handling can take place.
Thanks for pointing out the missing test, I forgot to cover this